### DISCUSSION TITLE
Mermaid rendering to docx halts indefinitely when run through command prompt or RStudio

### DISCUSSION THREAD

**Discussion by Gewerd-Strauss** [quarto-dev/quarto-cli#6085] [Q&A]
### Description

Hello,

I am experiencing a weird issue where converting a `.qmd`-doc with mermaid syntax to `docx` will halt code execution indefinitely by not closing the launched-therefore chrome.exe-instances required for mermaid rendering. From an automation perspective, this is obviously really bad.

---

# Files and commands issued for a MWE

The document used for testing purposes:

````qmd
---
creation date: 2023-06-30 07:26
modification date: 2023-07-01 12:24
tags:
- vault
---

```{mermaid}
    stateDiagram-v2
            [*] --> Still
            Still --> [*]
        
            Still --> Moving
            Moving --> Still
            Moving --> Crash
            Crash -->  [*] JUDAS 
    
```


```{mermaid}
    flowchart LR
      A[Hard edge] --> B(Round edge)
      B --> C{Decision}
      C --> D[Result one]
      C --> E[Result two]
```

```{mermaid}
    sequenceDiagram
      participant Alice
      participant Bob
      Alice->>John: Hello John, how are you?
      loop Healthcheck
        John->>John: Fight against hypochondria
      end
      Note right of John: Rational thoughts <br/>prevail!
      John-->>Alice: Great!
      John->>Bob: How about you?
      Bob-->>John: Jolly good!
```

````

This doc is converted using the following `build.R`-script[^2]:

````R
getwd()
if (getwd() != "C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground")
{
setwd("C:\\Users\\Claudius Main\\Desktop\\TempTemporal\\Mermaid Playground")
getwd()
}
getwd()
sprintf('Chosen Output formats:')
sprintf('quarto::html')
sprintf('quarto::docx')
version



yaml::write_yaml(list(df_print = "kable",
'fig-dpi' = 96,
'fig-height' = 6,
'fig-width' = 8,
'number-depth' = 99,
'number-sections' = TRUE,
'reference-doc' = "D:/Dokumente neu/PaperStyle_WordReferenceFiles/BE28 Template Internship Report - Kopie.docx",
toc = TRUE,
'toc-depth' = 5,
'toc-title' = "Table of Contents"),"C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_docx.yaml")

yaml_content <- readLines("C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_docx.yaml")
yaml_content <- stringr::str_replace(yaml_content,": no\"",": FALSE\"")
yaml_content <- stringr::str_replace(yaml_content,": yes\"",": TRUE\"")
yaml_content <- stringr::str_replace(yaml_content,": yes",": TRUE")
writeLines(yaml_content,"C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_docx.yaml")


yaml::write_yaml(list(df_print = "kable",
'email-obfuscation' = "none",
'fig-cap-location' = "margin",
'fig-height' = 6,
'fig-title' = "Figure",
'fig-width' = 8,
'number-depth' = 99,
'number-sections' = TRUE,
'tbl-cap-location' = "margin",
toc = TRUE,
'toc-depth' = 5,
'toc-location' = "right",
'toc-title' = "Table of Contents"),"C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_html.yaml")

yaml_content <- readLines("C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_html.yaml")
yaml_content <- stringr::str_replace(yaml_content,": no\"",": FALSE\"")
yaml_content <- stringr::str_replace(yaml_content,": yes\"",": TRUE\"")
yaml_content <- stringr::str_replace(yaml_content,": yes",": TRUE")
writeLines(yaml_content,"C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_html.yaml")

quarto::quarto_render("index.qmd",pandoc_args = c("--metadata-file","C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_docx.yaml"),output_format = "docx",output_file = "Mermaid Playground (quarto).docx")
sprintf('Finished quarto_render for docx')
quarto::quarto_render("index.qmd",pandoc_args = c("--metadata-file","C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground/yaml_html.yaml"),output_format = "html",output_file = "Mermaid Playground (quarto).html")
print(all_times)

````

And finally, this is the command I feed[^1] to `cmd.exe` kicking the whole thing off:

```cmd
"<PathToRScript.exe>" "<PathToBuildR.R>"
"C:\Users\Claudius Main\AppData\Local\Programs\R\R-4.2.3\bin\x64\Rscript.exe" "C:\\Users\\Claudius Main\\Desktop\\TempTemporal\\Mermaid Playground\\build.R"
```
---

# Background

The `build.R`-script and the command fed to the cmd prompt are auto-generated by a [self-written autohotkey script](github.com/Gewerd-Strauss/ObsidianKnittr/)[^3][^4]. This procedure has been working flawlessly for rmarkdown and bookdown packages called by running the RScript shown above from the command line.

I have been recently made aware of the existence of quarto, and have since been implementing it into my script. While playing around with the features Quarto has over RMarkdown this issues occured.

QMD-documents not containing Mermaid syntax will get rendered completely successfully for both docx and html. 
I did not yet test for PDF (mostly because I just don't need pdf-outputs in 99.5% of my life, so no real priority in my limited time).

---

Pre-execution shows no chrome-processes running via System Informer:
![grafik](https://github.com/quarto-dev/quarto-cli/assets/80548781/cd020798-5bf3-4ce4-bff6-7212e8418151)

When the docx-file gets generated, three instances of chrome.exe get launched:
![grafik](https://github.com/quarto-dev/quarto-cli/assets/80548781/bc352272-2f00-4199-bb2c-f4f8683631d0)

Of those three chrome instances, two will have the same parent PID, and one will not. Forcefully killing the process with a unique parent PID will result in the script execution to continue, therefore generating the following codelines and giving the other intended document outputs.

Killing either of the two chrome instances with identical PID will not do anything.
Killing the unique PID will resolve the issue. It will also kill the remaining two processes automatically.
Not killing any processes and rerunning the build.R script after the docx-file is generated will again result in the docx-file to be generated, and launch an additional 3 chrome.exe processes. As before, there will be one with a unique parent PID. As before, killing that will result in this running instance to continue running.

So... I got little idea what is actually going on here, or why this happens. 

Any help and suggestions are welcome. 
I'll happily provide more information if required. 

---

# Info dump

R version as given by `version` via the build.R-script executed:

                   _                                
    platform       x86_64-w64-mingw32               
    arch           x86_64                           
    os             mingw32                          
    crt            ucrt                             
    system         x86_64, mingw32                  
    status                                          
    major          4                                
    minor          2.3                              
    year           2023                             
    month          03                               
    day            15                               
    svn rev        83980                            
    language       R                                
    version.string R version 4.2.3 (2023-03-15 ucrt)
    nickname       Shortstop Beagle                 

Complete return given by the execution of `build.R`:

```R
[1] "C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground"
[1] "C:/Users/Claudius Main/Desktop/TempTemporal/Mermaid Playground"
[1] "Chosen Output formats:"
[1] "quarto::html"
[1] "quarto::docx"
               _                                
platform       x86_64-w64-mingw32               
arch           x86_64                           
os             mingw32                          
crt            ucrt                             
system         x86_64, mingw32                  
status                                          
major          4                                
minor          2.3                              
year           2023                             
month          03                               
day            15                               
svn rev        83980                            
language       R                                
version.string R version 4.2.3 (2023-03-15 ucrt)
nickname       Shortstop Beagle                 
pandoc --output Mermaid Playground (quarto).docx
  to: docx
  default-image-extension: png
  number-sections: true
  reference-doc: >-
    D:/Dokumente neu/PaperStyle_WordReferenceFiles/BE28 Temples/BE28 Template Internship Report
    - Kopie.docx
  toc: true
  toc-depth: 5
  
metadata
  df_print: kable
  number-depth: 99
  toc-title: Table of Contents
  fig-dpi: 96
  fig-height: 6
  fig-width: 8
  number-sections: true
  reference-doc: 'D:/Dokumente neu/PaperStyle_WordReferenceFiles/BE28 Template Internship Report - Kopie.docx'
  toc: true
  toc-depth: 5
  
Output created: Mermaid Playground (quarto).docx

pandoc --output Mermaid Playground (quarto).html
  to: html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  email-obfuscation: none
  number-sections: true
  toc: true
  toc-depth: 5
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  df_print: kable
  fig-cap-location: margin
  fig-title: Figure
  number-depth: 99
  tbl-cap-location: margin
  toc-location: right
  toc-title: Table of Contents
  email-obfuscation: none
  fig-height: 6
  fig-width: 8
  number-sections: true
  toc: true
  toc-depth: 5
  
Output created: Mermaid Playground (quarto).html

Fehler in print(all_times) : Objekt 'all_times' nicht gefunden 
Ausführung angehalten
```

Return of `quarto check` via command prompt:

```cmd
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.3.433
      Path: C:\Users\Claudius Main\AppData\Local\Programs\Quarto\bin
      CodePage: 1252

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.11.1
      Path: C:/Users/Claudius Main/AppData/Local/Programs/Python/Python311/python.exe
      Jupyter: (None)

      Jupyter is not available in this Python installation.
      Install with py -m pip install jupyter

[>] Checking R installation...........OK
      Version: 4.2.3
      Path: C:/Users/CLAUDI~1/AppData/Local/Programs/R/R-42~1.3
      LibPaths:
        - C:/Users/Claudius Main/AppData/Local/Programs/R/R-4.2.3/library
      knitr: 1.43
      rmarkdown: 2.22

[>] Checking Knitr engine render......OK
```

---

[^1]: Sidenote: When automated via my script, the function [`GetStdStreams_WithInput()`](https://github.com/Gewerd-Strauss/ObsidianKnittr/blob/feat/quarto-implement-mermaidexecution/Lib/GetStdStreams_WithInput.ahk) will execute the command directly, without going through cmd.exe itself. Does not yield different results.

[^2]: Note that the final line throwing an error on`print(all_times)` is intentional in this case. It has nothing to do with the issue, and removing the infringing line will not yield different results.
[^3]: AHK is windows-only
[^4]: Specifically, [this branch](https://github.com/Gewerd-Strauss/ObsidianKnittr/tree/feat/quarto-implement-diagrams)

**Comment by Gewerd-Strauss** [comment]
Addendum:

Graphviz diagrams via 

````qmd
---
creation date: 2023-07-02 09:26
modification date: 2023-07-02 09:26
tags:
- vault
---

```{dot}
digraph G {
  start [shape=Mdiamond,color=red,bgcolor=red,style=filled];
  subgraph cluster_0 {
    color=red
    bgcolor=violet
    node [style=filled,color=white];
    a0 -> a1 -> a2 -> a3;
    label = "process #1";
  }

  subgraph cluster_1 {
    node [style=filled];
    b0 -> b1 -> b2 -> b3;
    b0 -> a3
    label = "process #2";
    color=yellow
    bgcolor="red"
  }
  end [shape=Msquare,color=green,style=filled];
  
  
  start -> a0;
  start -> b0;
  a1 -> b3;
  b2 -> a3;
  a3 -> b0
  a3 -> a0;
  a3 -> end;
  b3 -> end;
    
}
```

````

exhibits the same problem, launching 3 chrome instances which must be killed for code execution to resume.

**Comment by mcanouil** [comment]
I believe this to be an issue with using Chrome to convert the diagrams into static images.
Unfortunately I don't have the setup to fully check this.

FYI:

- #4359

**Reply by Gewerd-Strauss** [reply]
Well, I suppose I'll just have to find a way to auto-kill this process after X seconds then until this is resolved. Do we have any idea where the progress is in that regard?

**Reply by cscheid** [reply]
The problem I'm having with your issue is that it's very hard to reproduce given the different steps and scripts required. If you can produce a single self-contained .qmd file that causes the problem with a single shell command, then we can track this further.

**Comment by NixImagery** [comment]
This behaviour seems to be triggered by Mermaid diagrams in book projects, not specifically those also rendering docx.

When it behaves itself, during the (quarto) render task for book and pdf output, it uses chrome to render the mermaid diagram (I believe):

<img width="658" alt="Screenshot 2024-02-02 at 07 16 52" src="https://github.com/quarto-dev/quarto-cli/assets/5462406/635a9ed8-7799-48bc-8b0c-406db324b144">

These three vanish as the rendering completes, if there are no errors. Adding docx output adds an extra instance each of the three chrome processes, which also vanish on completion.

When the problem does occur, the `quarto render` gets stuck in the `.qmd` file containing the first mermaid script. A check of the Activity monitor shows that there are usually six or seven chrome-helper processes running, even if I haven't opened Chrome browser. I suspect this issue is related to the chrome helper process intermittently not shutting down nicely -- I am in the habit of expecting this  behaviour from time to time, and just killing anything "chrome" in Activity Monitor. It all works fine again, until it doesn't (maybe six or seven `quarto render`s later). 

Not a solution, I'm afraid, but at least I can keep moving forward when this happens. I hope that helps anyone else who gets stuck.

**Reply by mcanouil** [reply]
Thanks for this.

FYI, Chrome is used to get a raster image (PNG) of the Mermaid diagram since Word does not support Mermaid directly.

**Reply by NixImagery** [reply]
Thanks. I wonder why the behaviour also occurs occasionally in quarto projects I have which only render book and pdf. The trouble is, I also have some which do include docx, so I haven't managed to isolate a replicable case. I am just glad to have a workaround.

**Reply by mcanouil** [reply]
PDF (LaTeX) is similar to docx (Word) in regards to mermaid diagrams.
